packages/extension-commerce-bm-seo/src/components/seo-hoc.tsx

```diff
@@ -0,0 +1,128 @@
+
+        const {refetch} = useUrlMapping(
+            {
+                parameters: {
+                    urlSegment: urlSegment,
+                    locale: siteLocale
+                }
+            },
+            {
+                enabled: false
+            }
+        )
+
+        useEffect(() => {
+            const fetchData = async () => {
+                if (urlSegment) {
+                    const result = await refetch()
+                    if (resolveRef.current) {
+                        if (result.status === 'error') {
+                            resolveRef.current(undefined)
+                        } else {
+                            if (result.data?.destinationUrl) {
+                                resolveRef.current(result.data)
+                            } else {
+                                resolveRef.current(undefined)
+                            }
+                        }
+                    }
+                }
+            }
+            void fetchData().catch(console.error)
+        }, [urlSegment])
+
+        const {isBlocked: isNavigationBlocked} = useBlockNavigation(
+            async (location: Location, _: string) => {
+                const urlMappingResponse = await new Promise<UrlMappingResponse | undefined>(
+                    (resolve, __) => {
+                        const nextSegment = location.pathname
+                        // So that this promise can be resolved and navigation is unblocked outside this function
+                        resolveRef.current = resolve
+                        setUrlSegment(nextSegment)
+                    }
+                )
+                // If no redirect rule exists, go to original url
+                if (urlMappingResponse === undefined) {
+                    return
+                }
+                let Component
+                let props
+                // If the Redirect type is URL do a Redirect, else load matching component
+                if (!urlMappingResponse.resourceType) {
+                    Component = Redirect
+                    props = {
+                        to: urlMappingResponse.destinationUrl
+                    }
+                } else {
+                    Component = getComponent(
+                        routes,
+                        resourceTypeToComponentMap,
+                        urlMappingResponse.resourceType
+                    )
+                    props = {
+                        [`${urlMappingResponse.resourceType}Id`]: urlMappingResponse.resourceId
+                    }
+                }
+                setRoutes([
+                    {
+                        path: location.pathname,
+                        component: () => <Component {...props} />
+                    },
+                    ...routes
+                ])
+            }
+        )
+
+        useEffect(() => {
+            setIsNavigationBlocked(isNavigationBlocked)
+        }, [isNavigationBlocked])
```

Comment:
Similarly, we can add a comment here to note that we're broadcasting to other components via this shared state.